src/Parser/Core/Modules/RegenResourceCapTracker.js

```diff
@@ -0,0 +1,607 @@
+   * Calculate current maximum value for the resource. Accounting for items, talents, and active buffs.
+   * When checking combant's buffs/debuffs you may want to use this.combatantHasBuffActive() rather than
+   * the .hasBuff of the combatant directly, doing so will exclude buffs which wore off at this timestamp.
+   * @returns {number}  Maximum resource value, or null to try to use value from resource events.
+   */
+  currentMaxResource() {
+   /**
+    * If left unimplemented and no events provide max resource information the Analyzer will never know
+    * the resource's max value. It handles that situation without crashing, but without a cap value it
+    * is largely useless.
+    */
+    return null;
+  }
+
+  /**
+   * Applies an energize of the tracked resource type at the current timestamp.
+   * Use when a spec has a source of resource which doesn't appear as an energize event in the log.
+   * If you have an invisible energize that happens over a duration you should generate an event for
+   * each tick, as applying it all at once can lead to inaccuracies. If it doesn't have ticks consider
+   * handling it as a buff to the natural regeneration rate instead.
+   * @param {number} amount Raw amount of resources to gain. Limiting to max is done within this function
+   * so there's no need to check before passing it in.
+   */
+  processInvisibleEnergize(amount) {
+    if (amount == null || isNaN(amount)) {
+      throw new Error(`processInvisibleEnergize called without required parameter. amount: ${amount}`);
+    }
+    if (amount === 0) {
+      return;
+    }
+    const before = this.current;
+    const after = Math.min(before + amount, this.currentMaxResourceWithFallback());
+    const gain = after - before;
+    this.applyEnergize(gain);
+  }
+
+  /**
+   * Fetches the appropriate classResources object from an event, or null if there is none available.
+   * @param {object} event An event object which may have an appropriate .classResources property.
+   */
+  getResource(event) {
+    if (!event || !event.classResources) {
+      return null;
+    }
+    return event.classResources.find(r => r.type === this.constructor.resourceType.id);
+  }
+
+  /**
+   * A variation of this.selectedCombatant.hasBuff that excludes any buffs which were removed on the timestamp.
+   * @param {number} buffId ID of buff or debuff to check for on the current combatant.
+   * @param {number} timestamp Time to check, or null to use current timestamp. Cannot be a future timestamp.
+   */
+  combatantHasBuffActive(buffId, timestamp = null) {
+    if (!buffId || isNaN(buffId)) {
+      throw new Error(`combatantHasBuffActive called without required parameter. buffId: ${buffId}`);
+    }
+    if (!timestamp) {
+      timestamp = this.owner.currentTimestamp;
+    }
+    const buffHistory = this.selectedCombatant.getBuffHistory(buffId);
+    return !!buffHistory.find(buff => (buff.start <= timestamp && (!buff.end || buff.end > timestamp)));
```

Comment:
Afaik buffs are only removed from the history after the removebuff event, so the default behavior should already match the event order exactly.